tcp_sink.go

```diff
@@ -166,51 +171,78 @@ func (s *tcpStatsdSink) FlushTimer(name string, value float64) {
 }
 
 func (s *tcpStatsdSink) run() {
-	settings := GetSettings()
+	conf := GetSettings()
+	addr := net.JoinHostPort(conf.StatsdHost, strconv.Itoa(conf.StatsdPort))
+
 	t := time.NewTicker(flushInterval)
 	defer t.Stop()
 	for {
 		if s.conn == nil {
-			conn, err := net.Dial("tcp", fmt.Sprintf("%s:%d", settings.StatsdHost,
-				settings.StatsdPort))
-			if err != nil {
+			if err := s.connect(addr); err != nil {
 				logger.Warnf("statsd connection error: %s", err)
+
+				// TODO (CEV): don't sleep on the first retry
 				time.Sleep(3 * time.Second)
 				continue
 			}
-			s.conn = conn
 		}
 
 		select {
 		case <-t.C:
 			s.flush()
-		case buf, ok := <-s.outc: // Receive from the channel and check if the channel has been closed
-			if !ok {
-				logger.Warnf("Closing statsd client")
-				s.conn.Close()
-				return
-			}
-			lenbuf := len(buf.Bytes())
-			_, err := buf.WriteTo(s.conn)
-			if len(s.outc) == 0 {
-				// We've at least tried to write all the data we have. Wake up anyone waiting on flush.
-				s.mu.Lock()
-				s.lastFlushTime = time.Now()
-				s.mu.Unlock()
-				s.flushCond.Broadcast()
-			}
-			if err != nil {
-				s.mu.Lock()
-				s.handleFlushError(err, lenbuf)
-				s.mu.Unlock()
-				_ = s.conn.Close() // Ignore close failures
-				s.conn = nil
+		case <-s.doFlush:
+			// Only flush pending buffers, this prevents an issue where
+			// continuous writes prevent the flush loop from exiting.
+			//
+			// If there is an error writeBuffer() will set the conn to
+			// nil thus breaking the loop.
+			//
+			n := len(s.outc)
+			for i := 0; i < n && s.conn != nil; i++ {
+				buf := <-s.outc
+				s.writeBuffer(buf)
+				putBuffer(buf)
 			}
+			// Signal every blocked Flush() call. We don't handle multiple
+			// pending Flush() calls independently as we cannot allow this
+			// to block.  This is best effort only.
+			//
+			s.flushCond.Broadcast()
+		case buf := <-s.outc:
+			s.writeBuffer(buf)
 			putBuffer(buf)
 		}
 	}
 }
 
+// writeBuffer writes the buffer to the underlying conn.  May only be called
+// from run().
+func (s *tcpStatsdSink) writeBuffer(buf *bytes.Buffer) {
+	len := buf.Len()
+
+	// TODO (CEV): parameterize timeout
+	s.conn.SetWriteDeadline(time.Now().Add(defaultWriteTimeout))
+	_, err := buf.WriteTo(s.conn)
+	s.conn.SetWriteDeadline(time.Time{}) // clear
+
+	if err != nil {
+		s.mu.Lock()
+		s.handleFlushError(err, len)
+		s.mu.Unlock()
+		_ = s.conn.Close()
+		s.conn = nil // this will break the loop
```

Comment:
This seems unsafe to me.  I understand it's intended to hit https://github.com/lyft/gostats/pull/79/files#diff-fe0a6341412533cc5c3ed899e22b8610R201 to cause the loop to exit, but there is a race condition and e.g. https://github.com/lyft/gostats/pull/79/files#diff-fe0a6341412533cc5c3ed899e22b8610R224 could panic.

In general, there's a lack of synchronization around the creation of a new connection too (https://github.com/lyft/gostats/blob/ed6b3b5b8a51206060bbbb1804a81378cbb90a8b/tcp_sink.go#L249) which could lead to observing a partially initialized connection.

Reply:
Per slack this one is fine since run() is the only goroutine calling into here. I incorrectly thought there was concurrency.

Reply:
This is not a race since there is only ever 1 instance of the `run()` loop and it is the only code that calls this method.